FF.com containers docs updates (#11708)
microsoft/FluidFramework: docs/content/docs/build/containers.md

editor: tylerbutler
---

The container is the primary unit of encapsulation in the Fluid Framework.
It allows a group of clients to access the same set of shared objects and co-author changes on those objects.

> By the same token, it is a permission boundary ensuring visibility and access to a limited set of clients.

Reviewer: ```suggestion
It is also a permission boundary ensuring visibility and access only to permitted clients.
```

Your code creates containers using APIs provided by a service-specific client library.
Each service-specific client library implements a common API for manipulating containers.
For example, the [Tinylicious library]({{< relref "Tinylicious" >}}) provides [these APIs]({{< relref "docs/apis/tinylicious-client.md" >}}) for the Tinylicious Fluid service.

> These common APIs enable your code to specify what shared objects should live in the `FluidContainer`, and retreive `FluidContainer` once it is created.

Reviewer: ```suggestion
These common APIs enable your code to specify what shared objects should live in the `FluidContainer`, and retrieve the `FluidContainer` once it is created.
```

};
```

> Note above that even though `SharedString` was used in the `initialObject`, it should still be included in `dynamicObjectTypes` if you
> plan to create dynamic `SharedString` after the container has been created.

Reviewer: ```suggestion
Note that even though `SharedString` was used in the `initialObjects` in the example above, it must also be included in `dynamicObjectTypes` if you
want to create dynamic `SharedString` after the container has been created.
```

const containerId = await container.attach();
```

Notes:

- `client` represents an object defined by the service-specific client library. See the documentation for the service you are using for more details about how to use its service-specific client library.

> - It is a good practice to deconstruct the object that is returned by `createContainer` into its two main parts; `container` and `services`. For an example of the use of the latter, see [Working with the audience]({{< relref "audience.md#working-with-the-audience" >}}).

Reviewer: ```suggestion
- It is a good practice to destructure the object that is returned by `createContainer` into its two main parts; `container` and `services`. For an example using `services`, see [Working with the audience]({{< relref "audience.md#working-with-the-audience" >}}).
```

Destructuring seems to be the proper term: <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment>

> A newly created container is in a *detached* state. A detached container is not connected to the Fluid service and no data is shared with other clients.

Author: I tried to separate **attached** and **connected** terminology. This line also needs a change.

This is the point where you can create initial data to populate your shared objects if needed.

> This is often useful in the scenarios where you want to make sure that all connected clients have coherent initial state.

Reviewer: ```suggestion
This is often useful in scenarios where you want to make sure that all connected clients have a coherent initial state.
```

For example, this could mean setting a minimum table size if collaboration involves tables.

### Attaching a container

In order to attach the container to a service, call its `attach` method.

> Once *attached*, the Fluid container becomes a live entity living on Fluid service and can be loaded by other clients.

Reviewer: ```suggestion
Once *attached*, the Fluid container begins communicating with the Fluid service and can be loaded by other clients.
```

Is this clearer?

>     await client.getContainer(/*container id*/, schema);

Author: We are using container IDs in our telemetry to represent instance of the container runtime. Clients will often use our telemetry to log/observer errors so this will certainly create confusion. At the very least we likely need more explicit clarification on telemetry/logging docs.

```

### Deleting a container

Deleting a container is a service-specific feature, so you should refer to the documentation associated with the specific Fluid service you are using. See [Available Fluid Services]({{< relref "service-options.md" >}}) for more information about Fluid service options.

## Container lifecycle and manipulation

> ### attached/detached

Author: Entirely removed attached/detached section as it seems as repetition of earlier content in "creating" and "attaching" the container.

Newly created containers begin in a *detached* state.
A detached container is not connected to the Fluid service and no data is shared with other clients.
This is an ideal time to create initial data in your Fluid data model.

An *attached* container, conversely, **is** connected to the Fluid service, and data changes are shared with the other clients.

> An existing container being loaded by a client will always initially be in an attached state.

Reviewer: > An existing container being loaded by a client will always initially be in an attached state.

This seems worth preserving.

container.on("connected", () => {
    // enable editing if disabled.
});

// Hypothetical event firing when user becomes idle. It is for demonstration purposes only.
user.on("idle", () => {
    // Disconnect the container when idle to prevent unnecessary network traffic/COGS
    container.disconnect();
});

// Hypothetical event firing when user becomes active. It is for demonstration purposes only.
user.on("active", () => {
    // Connect the container when active to resume op processing
    container.connect();
});
```

Notes:

> - Connection is established by default whenever container is loaded or attached.

Reviewer: ```suggestion
- Connection is established by default whenever an existing container is loaded or a new container is attached.
```

"Loaded" isn't a state, right? This is just referring to existing vs. new containers?

> - You likely want to ensure that all pending changes are saved prior to calling `disconnect`.

Author: @scottn12 is this correct? Anything else to call out here or better communication connect/disconnect implications ?

Reviewer 2: This seems good to me! It might be worth mentioning that by default the container will try to reconnect automatically if connection is lost

See the following section on dirty/saved state for more details on that topic.

> ### Dirty/Saved state

Reviewer: ```suggestion
### Dirty/saved state
```

A container is considered **dirty** if it has local changes that have not yet been acknowledged by the service. You should always check the `isDirty` flag before disposing the container or disconneting from the service.

> Closing or disconnecting the container while `isDirty === true` may result in the loss of operations that have not yet been acknowledged by the service.

Reviewer: ```suggestion
If you close or disconnect the container while `isDirty === true`, you may lose operations that have not yet been acknowledged by the service.
```

  In some cases this can be due to lack of network connection.
  If the network connection is down, it needs to be restored for the pending changes to be acknowledged.

```typescript {linenos=inline}
if(container.isDirty) {
  container.on("saved", () => {
    // safe to dispose or disconnect the container
  });
}
```

> In terms of user experience, keep in mind that the `isDirty` flag will flicker as pending changes are awaiting acknowledgement.
> Therefore, the host may choose to incorporate some delay before interpreting the flag again, when basing the user experience off of its state.

Reviewer: ```suggestion
In terms of user experience, keep in mind that the `isDirty` value will change rapidly as pending changes are awaiting acknowledgement.
Therefore, the host may choose to incorporate some delay before reading the value again (debouncing), when basing the user experience off of its state.
```

Using the debouncing terms seems useful here.

The container emits the *saved* event to notify the caller that all the local changes have been acknowledged by the service and the document is marked as clean.

```typescript {linenos=inline}
container.on("saved", () => {
    // all pending edits have been saved at this point.
});
```

The container emits the *dirty* event to notify the caller that there are local changes that have not been acknowledged by the service yet and the document is still in a dirty state.

```typescript {linenos=inline}
container.on("dirty", () => {
    // container has pending changes that need to be ack-ed by the service.
});
```

### Disposing a container

The container object may be disposed to remove any server connections and clean up registered events. Once a container is disposed, your code must call `getContainer` again if it needs to  be reloaded.

```typescript {linenos=inline}
container.dispose();

const disposed = container.disposed;

container.on("disposed", () => {
    // handle event cleanup to prevent memory leaks
});
```

As mentioned above, you probably want to make sure all pending changes are saved prior to calling `dispose`.

### Deleting a container

> Deleting a container is a service-specific feature, so you should refer to the documentation associated with the specific Fluid service you are using. See [Available Fluid Services]({{< relref "service-options.md" >}}) for more information about Fluid service options.

Author: Why is this "service-specific" feature? We should elaborate this further. If we are implying we don't want to expose this feature through API we should elaborate on why-s. The way we phrase it here it sounds like there is API we could use, but its service specific, and user should go look for it. Then when you look at our client packages, none of them offer this capability. The intention may be to communicate something different here? AFR has this capability: https://docs.microsoft.com/en-us/azure/azure-fluid-relay/how-tos/container-deletion.

Reviewer: It's service-specific because we think that data management of that type belongs at the service level. For Azure, for example, there are lots of existing storage APIs that AFR is integrating with for managing the lifecycle of the container data (when you think of containers as documents/files). Azure customers expect to be able to manage their "Fluid data" in the same way as their other data.

Author: I had few questions on this topic, but we can separate them from this PR. This is good clarifications for what I needed here.
